clk-gateway/src/index.ts

```diff
@@ -0,0 +1,518 @@
+        throw new Error(`Batch ${batchNumber} is not committed`);
+      }
+
+      const batchInfo = await getBatchInfo(batchNumber);
+      const storageProof: StorageProof = {
+        account: proof.address,
+        key: proof.storageProof[0].key,
+        path: proof.storageProof[0].proof,
+        value: proof.storageProof[0].value,
+        index: proof.storageProof[0].index,
+        metadata: {
+          batchNumber: batchInfo.batchNumber,
+          indexRepeatedStorageChanges: batchInfo.indexRepeatedStorageChanges,
+          numberOfLayer1Txs: batchInfo.numberOfLayer1Txs,
+          priorityOperationsHash: batchInfo.priorityOperationsHash,
+          l2LogsTreeRoot: batchInfo.l2LogsTreeRoot,
+          timestamp: batchInfo.timestamp,
+          commitment: batchInfo.commitment,
+        },
+      };
+
+      const data = AbiCoder.defaultAbiCoder().encode(
+        [STORAGE_PROOF_TYPE],
+        [storageProof],
+      );
+
+      res.status(200).send({
+        data,
+      });
+    } catch (error) {
+      const errorMessage =
+        error instanceof Error ? error.message : String(error);
+      res.status(500).send({ error: errorMessage });
+    }
+  },
+);
+
+app.post(
+  "/registerL2",
+  [
+    body("name")
+      .isFQDN()
+      .withMessage("Name must be a fully qualified domain name")
+      .custom((name) => {
+        const [sub, domain, tld] = name.split(".");
+        if (domain !== cnsDomain || tld !== cnsTld) {
+          return false;
+        }
+        return true;
+      })
+      .withMessage("Invalid domain or tld"),
+    body("owner")
+      .isString()
+      .custom((owner) => {
+        return isAddress(owner);
+      })
+      .withMessage("Owner must be a valid Ethereum address"),
+  ],
+  async (req: Request, res: Response) => {
+    try {
+      const authHeader = req.headers.authorization;
+      if (!authHeader) {
+        res.status(401).json({ error: "Authorization header is missing" });
+        return;
+      }
+      const idToken = authHeader.split("Bearer ")[1];
+      if (!idToken) {
+        res.status(401).json({ error: "Bearer token is missing" });
+        return;
+      }
+      const decodedToken = await admin.auth().verifyIdToken(idToken, true);
+      if (!decodedToken.email_verified) {
+        res.status(403).json({ error: "Email not verified" });
+        return;
+      }
+      if (decodedToken.subDomain) {
+        res.status(403).json({
+          error: "One subdomain already claimed with this email address",
+        });
+        return;
+      }
+      await admin.auth().revokeRefreshTokens(decodedToken.uid);
+
+      const result = validationResult(req);
+      if (!result.isEmpty()) {
+        res.status(400).json(result.array());
+        return;
+      }
+      const data = matchedData(req);
+
+      const sub = data.name.split(".")[0];
+      const owner = getAddress(data.owner);
+
+      const response = await clickNameServiceContract.register(owner, sub);
```

Comment:
This is where we might want to do manual nonce management (future ticket?)

Reply:
No manual nonce management just buffering the request and batching them to reduce the likelihood of concurrent registers

Reply:
https://linear.app/nodle/issue/CHA-546